- Developer documentation (#1)
UBAutograding/devU-api: README.md

# Typescript Example Project

> This is an API for the v4 autograder. The main purpose of this project is to act as an inbetween for the containers doing the autograding, and the users submitting assignments.

Reviewer: ```suggestion
This is an API for the v4 autograder. The main purpose of this project is to act as an in-between for the containers doing the autograding, and the users submitting assignments.
```

## Project Dependencies (install these)

> - [Node](https://nodejs.org/en/) - A javascript interpereter to run JS outside of the browser

Reviewer: ```suggestion
- [Node](https://nodejs.org/en/) - A javascript interpreter to run JS outside of the browser
```

npm run typeorm -- migration:run
```

> Once you've got all the dependancies installed you can run the project via

Reviewer: ```suggestion
Once you've got all the dependencies installed you can run the project via
```

If you're working in vscode, a configuration has been included to allow for debugging. Open up the VS Code Run and Debug section and click `Debug API`.

### Convenient Devtools

The project has some nice devtools built into it that will make your life easier. Leaving these here to hopefully help make the engineering process a bit smoother as well as keep someone a bit more sane.

> `environment.ts`: This is a rollup file for all the environment variables that the project uses as well as their fallbacks if none are provided. What this means is you can overwrite any of there easilly **without having to change code** if you ever have a need to do so. Here's a commandline tip: you can set temparary environment variables on a per process basis by adding them before the command you want to run. Example: If we wish to set some value for `LOG_DB` temporarily, we can run the api like:

Reviewer: ```suggestion
`environment.ts`: This is a rollup file for all the environment variables that the project uses as well as their fallbacks if none are provided. What this means is you can overwrite any of there easily **without having to change code** if you ever have a need to do so. Here's a command-line tip: you can set temporary environment variables on a per process basis by adding them before the command you want to run. Example: If we wish to set some value for `LOG_DB` temporarily, we can run the api like:
```

```

> LOG_DB=1 npm start # this is a helpfull command if you wish to log all sql commands to the DB when developing

Reviewer: ```suggestion
LOG_DB=1 npm start # this is a helpful command if you wish to log all sql commands to the DB when developing
```

[Prettier](https://prettier.io/) is a library that's already installed and deals with code formatting. It will format all of your code on commit so that you never have to worry about style matching to the project. If you wish to see your code styled in real time, VSCode has a prettier plugin that will format on save for you (you don't _have_ to do this, it'll still format on commit). Alternatively, you can manually format all of the code in the project with:

```
npm run format
```

Use [Azure Data Studio](https://docs.microsoft.com/en-us/sql/azure-data-studio/download-azure-data-studio?view=sql-server-ver15) (or some other sql client) for easier sql queries/ testing while developing. We're using [PostgresSQL](https://www.postgresql.org/) so keep that in mind while googling how to query things as the syntax and features can be slightly different across sql flavors. **[AZURE DATA STUDIO REQUIRES A POSTGRES PLUGIN TO WORK WITH POSTGRESQL](https://docs.microsoft.com/en-us/sql/azure-data-studio/extensions/postgres-extension?view=sql-server-ver15)**. So if you want to use Azure Data Studio, you'll have to install that extension.

## Working in the Project

> This projet is built using

Reviewer: ```suggestion
This project is built using
```

- [Express](https://expressjs.com/) - A Node REST framework.
- [TypeORM](https://typeorm.io/#/) - A Typescript ORM (manages the DB and all DB calls)

For those unfamiliar with Express, I'll attempt to give a brief rundown before talking about the project structure as a whole.

### Middleware

Express is a REST framework for node. Devs new to express will likely find one of the most perplexing parts of Express to be how it handles middleware. For those unfamiliar with what middleware is, it's largely a catch all term for code that connects together pieces of code in the api. In Express's context, middleware is basically everything that runs within each router; here's an example:

```typescript
Router.get('/:id', idAsInt, UserController.detail, serializer)
```

> In express, all of the functions added after the route's path are considered middleware and each route can have as many middleware as is needed. In this project we separate out `controllers` into thier own directory though they are still considered middleware.

Reviewer: ```suggestion
In express, all of the functions added after the route's path are considered middleware and each route can have as many middleware as is needed. In this project we separate out `controllers` into their own directory though they are still considered middleware.
```

> Express's middlware is always a function with three arguments

Reviewer: ```suggestion
Express's middleware is always a function with three arguments
```

```typescript
function middlwareName(req: Reques, res: Response, next: NextFunction)
```

Here's what you need to know:

- Middlewares are called sequentially in the order of the arguments list
- You pass data between middlewares by adding whatever you want to pass to the `req` object.
  - es `req.users = userResponse`
  - Because we're in typescript you'll have to add whatever parameter you're passing to `@types/express/index.d.s`
- Calling `next()` moves you to the next piece of middleware in the chain

>   - Caling `next` with _any_ arguments is considered throwing an error and will stop you from going to the next middleware, and it will bubble back the chain to the `globalErrorHandler`.

Reviewer: ```suggestion
  - Calling `next` with _any_ arguments is considered throwing an error and will stop you from going to the next middleware, and it will bubble back the chain to the `globalErrorHandler`.
```

### Control Flow

Here's the basic layout of the application

![control flow of the api](/docs/controlFlow.png 'Control Flow')

Let's take this from the top

- `index.ts`: where the application is bootstrapped from, controls all global server controls and middlewares
- `routers/index.ts`: largely a rollup for all the other routers. Can be used to add router specific middleware to routes/ subroutes

> - `routers/route.ts`: Individual routes for each resource, where the list of middlwares can be found. _All routers call unique middlware_

Reviewer: ```suggestion
- `routers/route.ts`: Individual routes for each resource, where the list of middleware can be found. _All routers call unique middleware_
```
